src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -1281,30 +1283,23 @@ private static IEnumerable<Diagnostic> GetEffectiveDiagnosticsImpl(ImmutableArra
                 yield break;
             }
 
+            if (compilation.SemanticModelProvider == null)
```

Comment:
Accepting any model here seems like it could easily lead to regressions in the future. Consider a few months from now if another part of the stack decides to add a light weight `SemanticModelProvider` for whatever reason. That would silently cause this part of the stack to regress behavior because it would cease to add `CachingSemanticModelProvider` instances. 

Did you consider enforcing that this requires having `CachingSemanticModelProvider`?

Reply:
This is an existing public API, so we can't enforce this unfortunately. This is just a best effort optimization when this method is invoked with large number of diagnostics in same file.